### PR TITLE
Basic Multiplayer Functionality

### PR DESCRIPTION
**Pull request by pastaq**
Stable development branch of Tuxemon multiplayer basic functionality. Players can do the following:
1.) Host a local game
2.) Join a local game
3.) Move around within the world and remain synced with all other players
4.) Run
5.) Switch maps and remain synced with all other players
6.) "I got 99 commits and a pull ain't one"

### CODE REVIEW

File: tuxemon/core/components/event/actions/combat.py [hunk 1]
```diff
@@ -250,4 +279,23 @@ def random_encounter(self, game, action):
             game.state_dict["WORLD"].menu_blocking = True
             player1.moving = False
             player1.direction = {'down': False, 'left': False, 'right': False, 'up': False}
+    
+    
+    def check_battle_legal(self, player):
```
Comment (pastaq) [inline tuxemon/core/components/event/actions/combat.py]: This code was used multiple places. Simplified into a single function.

File: tuxemon/core/components/event/actions/player.py [hunk 2]
```diff
@@ -120,11 +120,18 @@ def teleport(self, game, action):
                                     layer_pos += 1
                             y_pos += 1
                         x_pos += 1
-
+        
+        # Update the server/clients of our new map and populate any other players.
+        if game.isclient:
+            game.add_clients_to_map(game.client.client.registry)
+            game.client.update_player(player.facing)
```
Comment (pastaq) [inline tuxemon/core/components/event/actions/player.py]: Lines 128 -131 should be removed.

File: tuxemon/core/components/event/actions/player.py [hunk 3]
```diff
@@ -120,11 +120,18 @@ def teleport(self, game, action):
                                     layer_pos += 1
                             y_pos += 1
                         x_pos += 1
-
+        
+        # Update the server/clients of our new map and populate any other players.
```
Comment (pastaq) [inline tuxemon/core/components/event/actions/player.py]: line 125 should be: 
if game.isclient or game.ishost:

**Comment by pastaq** [comment]
Pull request adds two dependencies
netifaces module, sudo apt-get install python-netifaces
neteria, https://github.com/ShadowBlip/Neteria

File: .settings/org.eclipse.core.resources.prefs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+eclipse.preferences.version=1
```
Comment (ShadowApex) [inline .settings/org.eclipse.core.resources.prefs]: Please do not include IDE preferences in pull requests.

File: tuxemon/core/components/event/actions/combat.py [hunk 5]
```diff
@@ -26,6 +26,7 @@
 
 import logging
 import random
+from __builtin__ import False
```
Comment (ShadowApex) [inline tuxemon/core/components/event/actions/combat.py]: Why is this import needed?

File: tuxemon/core/components/menu/__init__.py [hunk 6]
```diff
@@ -988,14 +989,15 @@ def draw_textItem(self, textlist, columns=1, pos_x=0, pos_y=0, align="middle", a
             self.menudis_y += line_spacing + longest_item.get_height()
 
 
-    def get_event(self, event, game=None):
```
Comment (ShadowApex) [inline tuxemon/core/components/menu/__init__.py]: Why is the menu prototype `get_event` method changed to `menu_event`?

File: tuxemon/core/components/networking.py [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+            interface = netifaces.ifaddresses(device)
+            try:
+                self.interfaces[device] = interface[netifaces.AF_INET][0]['addr']
+            except KeyError:
+                pass
+        
+        for interface in self.interfaces:
+                        ip = self.interfaces[interface]
+                        if ip == "127.0.0.1": continue
+                        else: self.ips.append(ip)
```
Comment (ShadowApex) [inline tuxemon/core/components/networking.py]: So much indentation here!

File: tuxemon/core/components/networking.py [hunk 8]
```diff
@@ -0,0 +1,777 @@
...
+        :param None:
+
+        :rtype: None
+        :returns: None
+
+        """
+        events = []
+        for event_data in self.network_events:
+            if event_data == "KEYDOWN:up":
+                self.keys[pg.K_UP] = 1
```
Comment (ShadowApex) [inline tuxemon/core/components/networking.py]: Shouldn't `self.keys` be `self.game.keys` here? I don't see `self.keys` defined anywhere in TuxemonServer.

**Comment by ShadowApex** [comment]
The `README.md` file should also be updated to reflect the new dependencies introduced.

File: tuxemon/core/components/event/actions/combat.py (hunk 1 above)
Comment (ShadowApex) [inline tuxemon/core/components/event/actions/combat.py]: This is a good addition. Any redundant code should be made into a function for reuse-ability.

File: tuxemon/core/components/ai.py [hunk 9]
```diff
@@ -68,11 +68,50 @@ def make_decision(self, npc, opponent):
...
+        logger.info("Making descision")
+
+        #pprint.pprint(npc['monster'].__dict__)
+        #pprint.pprint(opponent)
+
+        return {'technique': 0}
+
+
+
+class PsudoAI(object):
```
Comment (ShadowApex) [inline tuxemon/core/components/ai.py]: Shouldn't this be "Pseudo"?

File: tuxemon/core/components/networking.py [hunk 10]
```diff
@@ -0,0 +1,777 @@
...
+        elif event_data["event_number"] <= self.server.registry[cuuid]["event_list"][event_data["type"]]:
+            return False
+        else:
+            self.server.registry[cuuid]["event_list"][event_data["type"]] = event_data["event_number"]
+                
+        if event_data["type"] == "PUSH_SELF":
+            self.server.registry[cuuid]["sprite_name"] = event_data["sprite_name"]
+            self.server.registry[cuuid]["map_name"] = event_data["map_name"]
+            self.server.registry[cuuid]["char_dict"] = event_data["char_dict"]
+            self.notify_populate_client(cuuid, event_data)
```
Comment (ShadowApex) [inline tuxemon/core/components/networking.py]: Note that this way of pushing a player's character data will make it very easy for a user to change their player data into whatever they want (e.g. All level 100 monsters, max stats, etc.). In the future, we'll need to have a standalone server that will store every player's position, monsters, stats, etc.

File: .settings/org.eclipse.core.resources.prefs (hunk 4 above)
Comment (pastaq) [inline .settings/org.eclipse.core.resources.prefs]: Woops. Will be removed.

File: tuxemon/core/components/ai.py (hunk 9 above)
Comment (pastaq) [inline tuxemon/core/components/ai.py]: Yes.

File: tuxemon/core/components/event/actions/combat.py (hunk 5 above)
Comment (pastaq) [inline tuxemon/core/components/event/actions/combat.py]: Will be removed.

File: tuxemon/core/components/menu/__init__.py (hunk 6 above)
Comment (pastaq) [inline tuxemon/core/components/menu/__init__.py]: Disambiguation. Multiple other classes use the name get_event for functions that retrieve conditional variables from a list or dict. Events are pushed to the menu class from a parent rather than retrieved by the menu's get_event (now menu_event). It can be changed back if necessary. The get_menu_event function in core/tools.py acts as the event retrieval and central source of control for menu's, eliminating redundant code present in each menu. Only new menu's have been modified to use this "menu controller" to ensure backwards compatibility and prevent the PR from going beyond the scope of it's intention. Each menu will now only contain use-case specific code in the menu_event function.

File: tuxemon/core/components/networking.py (hunk 7 above)
Comment (pastaq) [inline tuxemon/core/components/networking.py]: Good eye. I missed that. Will be fixed.

File: tuxemon/core/components/networking.py (hunk 10 above)
Comment (pastaq) [inline tuxemon/core/components/networking.py]: Agreed. This method of multiplayer would enable people to do multiplayer interaction and play with their local save game. Pushing this to a server will force you to make a new network only character. Once a network infrastructure is in place that can be targeted for development this will be implemented.

File: tuxemon/core/components/networking.py (hunk 8 above)
Comment (pastaq) [inline tuxemon/core/components/networking.py]: Fixed the issue and created a ControllerServer class that will start if net_controller_enabled = 1 in tuxemon.cfg. It uses the default port. The TuxemonServer has now been hard coded to 40081. The controller will need to be configured to only join a server with port 40080 or another port decided on later.
